core/src/autogluon/core/scheduler/remote/remote_manager.py

```diff
@@ -40,66 +42,72 @@ def __new__(cls):
 
     @classmethod
     def get_master_node(cls):
-        return cls.NODES[cls.MASTER_IP]
+        with read_lock(cls.LOCK):
+            ip = cls.NODES[cls.MASTER_IP]
+        return ip
 
     @classmethod
     def start_local_node(cls):
         port = cls.get_port_id()
         with warning_filter():
             remote = Remote(cls.MASTER_IP, port, local=True)
-        with cls.LOCK:
+        with write_lock(cls.LOCK):
             cls.NODES[cls.MASTER_IP] = remote
 
     @classmethod
     def launch_each(cls, launch_fn, *args, **kwargs):
-        for node in cls.NODES.values():
-            node.submit(launch_fn, *args, **kwargs)
+        with read_lock(cls.LOCK):
+            for node in cls.NODES.values():
+                node.submit(launch_fn, *args, **kwargs)
 
     @classmethod
     def get_remotes(cls):
-        return list(cls.NODES.values())
+        with read_lock(cls.LOCK):
+            nodes = list(cls.NODES.values())
+        return nodes
 
     @classmethod
     def upload_files(cls, files, **kwargs):
         if isinstance(files, str):
             files = [files]
-        for node in cls.NODES.values():
-            node.upload_files(files, **kwargs)
+        with read_lock(cls.LOCK):
+            for node in cls.NODES.values():
+                node.upload_files(files, **kwargs)
 
     @classmethod
     def add_remote_nodes(cls, ip_addrs):
-        ip_addrs = [ip_addrs] if isinstance(ip_addrs, str) else ip_addrs
-        remotes = []
-        for node_ip in ip_addrs:
-            if node_ip in cls.NODES.keys():
-                logger.warning('Already added remote {}'.format(node_ip))
-                continue
-            port = cls.get_port_id()
-            remote = Remote(node_ip, port)
-            with cls.LOCK:
+        with write_lock(cls.LOCK):
```

Comment:
Why are you wrapping read lock into the `write_lock` function? (I am unfamiliar with these functions)

Reply:
These are contexts. What it does: it acquires write lock, runs the function inside, then releases the lock. This is custom `RWLock`, which non-blocking for read operations, but will block read/write if write lock is acquired.